### DISCUSSION TITLE
[Compilers] `bytecodehash` property in metadata settings doesn't have `ipfs` option

### DISCUSSION THREAD

**Discussion by kumaryash90** [zkSync-Community-Hub/zksync-developers#194] [Compilers] [answered]
### Which compiler are you using?

- [X] zksolc
- [ ] zkvyper

### Which version of Solidity or Vyper?

0.8.17

### zksolc/zkvyper version?

1.3.15

### Explain the issue

While the latest zksolc compiler versions have a "metadata" option in settings, it doesn't allow setting it equal to "ipfs", which is usually a default setting for solc compiler.

i.e. it should be possible to set the `bytecodehash` below to `ipfs`

```bash
zksolc: {
    version: "1.3.15",
    compilerSource: "binary",
    settings: {
      metadata: {
        bytecodeHash: "none",
      },
      optimizer: {
        enabled: true,
      },
    },
  },
```

**Comment by idea404** [comment]
Hi @kumaryash90 

Thanks for bringing this up, our team is looking into this. Adding more hash formats should be possible there as long as they are padded to 256 bits.

**Reply by idea404** [reply]
Also @kumaryash90 to understand your request a little better: 

Given [the differences between `zkEVM` and `EVM`](https://era.zksync.io/docs/reference/architecture/differences-with-ethereum.html), our bytecode is different so there may be more inconsistencies than this value for `"metadata"`. Is this the single blocker for your project or would there be any other inconsistencies other than this config?

**Reply by idea404** [reply]
@kumaryash90 There will be a build with the ipfs patch, would you be able to test it?

**Reply by kumaryash90** [reply]
Thanks @idea404, I'll test the build.

For your previous question, so far this is the main blocker for integration with our deployment setup. We take into account the inconsistencies arising out of differences mentioned in the docs, however I'll reach out if we run into any other unnoticed issue.

**Reply by idea404** [reply]
Also @kumaryash90, this extra info would be very helpful: what is this used for in your project and what are the assumptions once this is supported? What do you expect/would like to happen?

**Reply by kumaryash90** [reply]
@idea404 we use the encoded ipfs hash for thirdweb's contract explorer

Basically when a contract is deployed via thirdweb CLI, the compiler metadata is uploaded to ipfs (this ipfs hash matches the one appended at the end of the bytecode) -- this allows thirdweb dashboard to read the bytecode, decode ipfs hash and pull the compiler metadata from ipfs to display the source code and abi etc, for e.g. https://thirdweb.com/mumbai/0xc15a838374d694E95a17C0F1BBe1b0d35EC2Fa4A

In case of zksync, right now we store the ipfs uri for each deployed contract in a registry since we can't decode it from the bytecode. Once the ipfs option is supported, we'd be able to display zksync deployed contracts on thirdweb dashboard like other chains, as well as simplify contract verification, without resorting to a registry.

**Reply by hedgar2017** [reply]
Hey @kumaryash90, we have a few more questions to clarify the mechanics of IPFS you'd like to rely on:
1. Do you need ipfs URLs in standard JSON output?
2. What/who is responsible for uploading the metadata to IPFS storage?
3. Do you depend on CBOR or literal content enabled by the fields below?
```
   "metadata": {
      // Reflects the setting used in the input json, defaults to "true"
      "appendCBOR": true,
      // Reflects the setting used in the input json, defaults to "ipfs"
      "bytecodeHash": "ipfs",
      // Reflects the setting used in the input json, defaults to "false"
      "useLiteralContent": true
    },
```
We can be handling it all somewhat differently. Anything can break your assumptions if you have any.

**Reply by kumaryash90** [reply]
@hedgar2017 
1. Yes ipfs URLs in standard JSON output would be needed
2. We upload the metadata to IPFS when deploying the contract (or while publishing on our website)
3. CBOR is how we do it right now

But let me know what will work and what won't. And how much this metadata setup will follow solc output.

**Reply by hedgar2017** [reply, marked as answer]
> @hedgar2017 
> 
> 1. Yes ipfs URLs in standard JSON output would be needed
> 
> 2. We upload the metadata to IPFS when deploying the contract (or while publishing on our website)
> 
> 3. CBOR is how we do it right now
> 
> 
> 
> But let me know what will work and what won't. And how much this metadata setup will follow solc output.

Thanks for the info.

In this case it will take a little longer, as:
1. We'll have to look at how solc generates those URLs and reverse-engineer in Rust.
2. Apparently, there is no minimal IPFS Rust library.
3. CBOR is doable, but we'll have to adjust the bytecode encoding so its size is a multiple of 32 after appending it.

**Reply by kumaryash90** [reply]
hey @hedgar2017 any updates on this?

**Reply by hedgar2017** [reply]
> hey @hedgar2017 any updates on this? 

Not yet.

This support turned to be non-trivial on our architecture. We'll try to find resources to investigate how to implement this properly.

**Reply by kumaryash90** [reply]
Thanks @hedgar2017. Appreciate it.
Please keep me informed about any progress or how I might assist on this.

**Reply by hedgar2017** [reply]
Hey @kumaryash90, would it work if we kept IPFS hashes in these URLs in metadata (e.g. `"dweb:/ipfs/QmSZ3HNGZom6N6eb8d74Y7UQAKAGRkXgbinwVVLaiuGb3S\"`), and appended the IPFS hash of the metadata to the bytecode in the same manner we're doing it with `keccak256`?

**Reply by kumaryash90** [reply]
@hedgar2017 yes that should work. By IPFS Hash you mean CBOR encoded data, right?

**Reply by hedgar2017** [reply]
@kumaryash90 unfortunately the solution above won't work as we need to rehash everything with our extra ZKsync-specific metadata anyway.
Talking of CBOR, do you prefer 32-byte zero-padding before the CBOR payload or after?

**Reply by kumaryash90** [reply]
Either is fine, before or after.

Btw, how will rehashing everything with zksync specific metadata change the CBOR encoded payload?

**Reply by hedgar2017** [reply]
@kumaryash90 that's another problem we have no clear answer for yet :(

**Reply by hedgar2017** [reply]
Hey @kumaryash90, could you try [this build](https://github.com/matter-labs/era-compiler-solidity/actions/runs/10432997241)?
To see the IPFS hash of the metadata, call `zksolc input.sol --metadata-hash ipfs`. It is also usable in standard JSON and through Hardhat as another hash type: `ipfs` | `keccak256` | `none`.
Please let me know if you need any of the following:
1. A separate `metadata` endpoint in `solc` CLI to see the metadata preimage JSON. Most of our users interact with `solc` through standard JSON, and it's already included there.
2. CBOR encoding of the IPFS hash, as now it is raw 34 bytes at the end, with a 30-byte padding before it.

**Reply by kumaryash90** [reply]
thanks @hedgar2017. checking it out today.

**Reply by hedgar2017** [reply]
Any updates @kumaryash90 ?

**Reply by kumaryash90** [reply]
@hedgar2017 sorry testing it now. 
What's the right way to download this build and use with my hardhat setup?

**Reply by hedgar2017** [reply]
> @hedgar2017 sorry testing it now. 
> 
> What's the right way to download this build and use with my hardhat setup?

For simplicity, I suggest simply replacing the zksolc Hardhat has downloaded into your system.

**Reply by kumaryash90** [reply]
> > @hedgar2017 sorry testing it now.
> > What's the right way to download this build and use with my hardhat setup?
> 
> For simplicity, I suggest simply replacing the zksolc Hardhat has downloaded into your system.

Replacing with? is there a build version i can specify?
Or just download and specify the path to this build?

**Reply by hedgar2017** [reply]
> > > @hedgar2017 sorry testing it now.
> 
> > > What's the right way to download this build and use with my hardhat setup?
> 
> > 
> 
> > For simplicity, I suggest simply replacing the zksolc Hardhat has downloaded into your system.
> 
> 
> 
> Replacing with? is there a build version i can specify?
> 
> Or just download and specify the path to this build?

I meant finding zksolc previously downloaded by Hardhat in your system and replacing it.
But you can probably use `binaryPath` or `compilerPath` setting in the Hardhat config. Not sure how is it called now.

**Reply by hedgar2017** [reply]
> > > @hedgar2017 sorry testing it now.
> 
> > > What's the right way to download this build and use with my hardhat setup?
> 
> > 
> 
> > For simplicity, I suggest simply replacing the zksolc Hardhat has downloaded into your system.
> 
> 
> 
> Replacing with? is there a build version i can specify?
> 
> Or just download and specify the path to this build?

I'm going to ask our Hardhat plugin devs to add HTTPS URLs there. Should be easier that way.

**Reply by hedgar2017** [reply]
Hey @kumaryash90, how are things?
Have you managed to try the build?
We'd like to include the feature you requested in the next release, so your help will be very appreciated.

**Reply by hedgar2017** [reply]
Just received the answer from Hardhat devs.
You can specify the zksolc executable URL like this:
```
zksolc: {
    compilerSource: 'binary',
    settings: {
        compilerPath: 'https://github.com/matter-labs/era-compiler-solidity/actions/runs/10432997241/artifacts/1823061347',
    }
}
```
Just pick the right URL for your platform.

**Reply by kumaryash90** [reply]
@hedgar2017 Just tried this build. 
I can now see the raw bytes of ipfs hash at the end by using the `--metadata-hash ipfs`  flag. 👍 

CBOR encoding would be preferred, to keep the decoding process same as other solc contracts. Along with last two bytes denoting the length of CBOR data.
The encoded data is an object containing "ipfs" and any other fields as described in solidity docs here: https://docs.soliditylang.org/en/latest/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode

e.g.
```
{
  "ipfs": "0x1220c019e4614043d8adc295c3046ba5142c603ab309adeef171f330c51c38f14989",
  ...
}
```

Let me know if that would be possible.

**Reply by kumaryash90** [reply]
Also, how is the current standard JSON different from the separate metadata endpoint's preimage JSON?

If this preimage JSON is the solc metadata associated with the ipfs hash generated above, then I think it will be required, so that we can upload this json at the expected ipfs location and decode/download using the bytecode when needed.

**Reply by kumaryash90** [reply]
@hedgar2017 Btw, the ipfs/metadata will also contain all the sources, right? along with both solc and zkversions?
Basically, what the standard solc flow does.

**Reply by hedgar2017** [reply]
Hey @kumaryash90, thank you for the detailed feedback. Let me walk through the key points.
1. `CBOR encoding would be preferred, to keep the decoding process same as other solc contracts. Along with last two bytes denoting the length of CBOR data.`
Yes, I can easily do that, but can we avoid adding compiler versions there? We have three of them: `solc` version, `solc` fork revision, `zksolc` version. It is too much to include as we'd like to keep the metadata size below 64 bytes. All this data is already included in the metadata file, so you can retrieve it from there.
2. `Also, how is the current standard JSON different from the separate metadata endpoint's preimage JSON?`
There is no such endpoint yet, but I can add it. The metadata JSON will contain the original `solc`'s metadata JSON, and additional fields like `zksolc` version and its optimized settings. So you'll always have everything `solc` provides, plus ZKsync's extra.
`Btw, the ipfs/metadata will also contain all the sources, right? along with both solc and zkversions? Basically, what the standard solc flow does.`
If it's provided by `solc`'s in its metadata, then yes. But I don't see it in the `solc` metadata format. It only provides IPFS hashes of source files. Is it what you're looking for?

**Reply by kumaryash90** [reply]
1. Yeah it's fine, we can just keep ipfs in the object.
2. Okay, that's what we need. Sounds good.
3. So, i was wondering, what does the ipfs hash point to? Is it the solc metadata for the target contract, or does it contain zksolc information as well. 
Also, the sources in metadata will have code/content or will they have ipfs metadata hashes as well?

**Reply by hedgar2017** [reply]
@kumaryash90 3.
The IPFS hash will be pointing to our updated metadata of course, which is a superset of `solc`'s metadata.
```
Also, the sources in metadata will have code/content or will they have ipfs metadata hashes as well?
```
As I can see, `solc` metadata does not contain sources, but it does contain IPFS hashes of sources, and since we're simply propagating wrapped `solc` metadata, those IPFS hashes are unchanged just like the sources themselves.
I'm almost done with it so will send you the new build shortly.

**Reply by kumaryash90** [reply]
Great. Thanks!

**Reply by hedgar2017** [reply]
@kumaryash90 please try [this build](https://github.com/matter-labs/era-compiler-solidity/actions/runs/10705654573)!

**Reply by kumaryash90** [reply]
Tested this @hedgar2017. Looks good. 

I couldn't use hardhat though. I'm assuming the hardhat plugin will need to be updated too, after this build is released?

**Reply by hedgar2017** [reply]
> Tested this @hedgar2017. Looks good. 
> 
> 
> 
> I couldn't use hardhat though. I'm assuming the hardhat plugin will need to be updated too, after this build is released?

Yeah, they'll update Hardhat.
What problem did you have exactly?

**Reply by kumaryash90** [reply]
`bytecodeHash: "ipfs"` is not recognized as valid.

also, there's some issue with using the downloaded binary in hardhat. it says the binary is corrupted (seems unlikely, i compared the hashes).

although, i was able to run it directly.

**Reply by hedgar2017** [reply]
@kumaryash90 I see.
Could you dump here full logs of both problems?
Did you rely on rosetta? Maybe you have Mac arm64 but downloaded a binary for amd64.
Are the Hardhat deps the latest version?

**Reply by kumaryash90** [reply]
Thanks @hedgar2017. It now works with hardhat. Tested the pre-release.

Btw, how did you encode the CBOR data? I'm having some issue with decoding it when using `cbor-x` library.

**Reply by kumaryash90** [reply]
@hedgar2017 Looks like the appended CBOR data has an issue in the first byte.

for e.g., in this 42 byte data:
```
a2646970667358221...ddb787bbd355c730f002a
```

The `a2` value at the beginning should instead be `a1` since there's just one key-value pair in the encoded cbor object, instead of two. It causes the the decoder to continue reading after the buffer has ended, expecting a second field in the data.

Can you take a look? 

Sorry I didn't catch it before since it worked with another tool I checked with (which perhaps works differently than other commonly used cbor libraries like cbor-x etc).

**Reply by hedgar2017** [reply]
@kumaryash90 indeed, thanks!
Will fix shortly!

**Reply by hedgar2017** [reply]
@kumaryash90 try [this](https://github.com/matter-labs/era-compiler-solidity/actions/runs/10858686422) please!

**Reply by kumaryash90** [reply]
thanks @hedgar2017, this worked!

**Reply by kumaryash90** [reply]
hey @hedgar2017, is this change available on the explorer verification api as well?

I'm getting a compilation error when sending verification request.

**Reply by hedgar2017** [reply]
> hey @hedgar2017, is this change available on the explorer verification api as well?
> 
> 
> 
> I'm getting a compilation error when sending verification request.

Please share all errors when making a report!
It will allow us to resolve issues more quickly.
Thanks!

**Reply by kumaryash90** [reply]
@hedgar2017 I'm sending a post request to this endpoint for my zksync-sepolia contract
- `https://block-explorer-api.sepolia.zksync.dev/api`
- contract: https://sepolia.explorer.zksync.io/address/0x0f7569568a1fbe9d4227a42059e32b0bb97f1995#contract

Our verification flow follows the official api documentation from zksync explorer here: https://block-explorer-api.sepolia.zksync.dev/docs#/Contract%20API/ApiController_verifyContractSourceCode

#### Error returned:
```
{ status: '0', message: 'NOTOK', result: 'Compilation error' }
```

This is only for contracts with latest changes wrt IPFS settings in zksolc.

Older compiled contracts are getting verified successfully.

**Reply by kumaryash90** [reply]
#### Update:

I was able to make it work by enforcing optimizer runs as follows:
```
settings: {
   ...
   optimizer: {
       enabled: true,
       runs: 200,
   },
}
```

But it doesn't work when I use the entire optimizer object provided in compiled metadata. Is the above setting a recommended one then, enforcing it to this value?

Also, there's still a couple of contracts that don't get verified. Here's a github gist containing the request body I'm sending:
https://gist.github.com/kumaryash90/e23ed3c376d53114889567d08d71d30e#file-zksync-contract-verification-request

Contract: https://sepolia.explorer.zksync.io/address/0xaaAF436511C2b5eC0ccABb061f9987432Eb64Ec2#contract

**Reply by kumaryash90** [reply]
There's another contract that gets doesn't get verified on zksync mainnet.

https://explorer.zksync.io/address/0x28bad9f3Ff6Bd5332BeE74EDf753C658DFA441fc#contract

But the api call returns a success message despite it:
```
{ status: '1', message: 'OK', result: 'Pass - Verified' }
```

Interestingly, this same contract at the same address is verified on zksync sepolia here:
https://sepolia.explorer.zksync.io/address/0x28bad9f3Ff6Bd5332BeE74EDf753C658DFA441fc

**Reply by vasyl-ivanchuk** [reply]
Hi @kumaryash90 , about the https://explorer.zksync.io/address/0x28bad9f3Ff6Bd5332BeE74EDf753C658DFA441fc#contract
we don't see any verification requests for this contract on Mainnet. Is there any chance you used the wrong endpoint or environment? Could you please try verifying it on Mainnet again?

What exactly are you using for the verification? Is it the Block Explorer API, and how are you initiating the request?

Thank you!

**Reply by vasyl-ivanchuk** [reply]
@kumaryash90 I managed to verify [0x28bad9f3Ff6Bd5332BeE74EDf753C658DFA441fc](https://explorer.zksync.io/address/0x28bad9f3Ff6Bd5332BeE74EDf753C658DFA441fc#contract) on Mainnet using its verified source code from Sepolia.

**Reply by kumaryash90** [reply]
thanks @vasyl-ivanchuk 

I was sending the verification request to this API endpoint for zksync mainnet explorer:
https://block-explorer-api.mainnet.zksync.io/api

Is this correct?

**Reply by vasyl-ivanchuk** [reply]
@kumaryash90 yes, it is correct.

**Comment by longbipubacr** [comment]
The zksolc compiler, version 1.3.15, does not currently support setting the metadata bytecodeHash option to "ipfs" directly. While it allows for specifying "none", setting it to "ipfs" is not yet supported in the current version. Consider reaching out to the maintainers of the zksolc compiler or checking for updates to see if this feature has been implemented in newer versions.

**Comment by npam90** [comment]
The latest zksolc compiler versions do not currently support setting the "metadata" option to "ipfs" for bytecode hash.

**Comment by deosovitis** [comment]
The user is reporting an issue with the zksolc compiler version 1.3.15, which does not allow the "metadata" option to be set to "ipfs" in settings. The user provides an example of the desired configuration, which includes setting the bytecodehash to "ipfs". No repository link was provided.
